fastfield_codecs/src/null_index/sparse.rs

```diff
@@ -0,0 +1,534 @@
+/// faster access.
+///
+/// The lower 16 bits of doc ids are stored as u16 while the upper 16 bits are given by the block
+/// id.
+///
+/// # Serialized Data Layout
+/// The data starts with the raw data of u16 encoded positions of the blocks. [u16 LE, .. #repeat*;
+/// Desc: Positions with values in a block]
+///
+/// The data is followed by block metadata, to know which area of the raw data belongs to which
+/// block. Only metadata for blocks with elements is recorded to
+/// keep the overhead low for scenarios with many very sparse columns. The block metadata consists
+/// of the block index and the number of values in the block.
+/// The last u16 is storing the number of metadata blocks.
+/// [u16 LE, .. #repeat*; Desc: Positions with values in a block][(u16 LE, u16 LE), .. #repeat*;
+/// Desc: (Block Id u16, Num Elements u16)][u16 LE; Desc: num blocks with values u16]
+///
+/// # Opening
+/// When opening the data layout, the data is expanded to `Vec<SparseCodecBlock>`, where the
+/// index is the block index.
+pub struct SparseCodec {
+    data: OwnedBytes,
+    blocks: Vec<SparseCodecBlock>,
+}
+
+/// A block consists of max u16 values
+#[derive(Debug)]
+struct SparseCodecBlock {
+    /// The number of values in the block
+    num_vals: u16,
+    /// The number of values set before the block
+    offset: u32,
+}
+
+impl SparseCodecBlock {
+    fn empty_block(offset: u32) -> Self {
+        Self {
+            num_vals: 0,
+            offset,
+        }
+    }
+
+    fn new(num_vals: u16, offset: u32) -> Self {
+        Self { num_vals, offset }
+    }
+
+    #[inline]
+    fn value_at_idx(&self, data: &[u8], idx: u16) -> u16 {
+        let start_offset: usize = (self.offset + idx as u32) as usize * 2;
+        let bytes: [u8; 2] = data[start_offset..start_offset + 2].try_into().unwrap();
+        u16::from_le_bytes(bytes)
+    }
+
+    #[inline]
+    #[allow(clippy::comparison_chain)]
+    // Looks for the element in the block. Returns the positions if found.
+    fn binary_search(&self, data: &[u8], target: u16) -> Option<u16> {
+        let mut size = self.num_vals;
+        let mut left = 0;
+        let mut right = size;
+        while left < right {
+            let mid = left + size / 2;
+
+            let mid_val = self.value_at_idx(data, mid);
+
+            if target > mid_val {
+                left = mid + 1;
+            } else if target < mid_val {
+                right = mid;
+            } else {
+                return Some(mid);
+            }
+
+            size = right - left;
+        }
+        None
+    }
+}
+
+const SERIALIZED_BLOCK_METADATA_SIZE: usize = 4;
+
+fn deserialize_sparse_codec_block(data: &[u8]) -> Vec<SparseCodecBlock> {
+    // The number of vals so far
+    let mut offset = 0;
+    let mut sparse_codec_blocks = Vec::new();
+    let num_blocks = u16::from_le_bytes([data[data.len() - 2], data[data.len() - 1]]);
```

Comment:
```suggestion
    let num_blocks = u16::from_le_bytes(&data[data.len() - 2..].try_into().unwrap());
```

Reply:
I would only use `unsafe` when we can see the difference in the benchmark

Reply:
Let's forget the unsafe part then, we already have a different  implementation of this in one place...
Can we extract it as a function mark it as inline and use it everywhere in this file.